camligorkem/assignment4/CountIslands.java

```diff
@@ -0,0 +1,126 @@
+import java.awt.Point;
+import java.util.HashSet;
+import java.util.Stack;
+import static java.lang.Math.abs;
+
+public class CountIslands {
+    /**
+     * Given a 2d map of boolean tiles (false = water, true= island), number of rows and number of cols
+     * returns the count of islands
+     * @param tilesMap 2d map of boolean tiles
+     * @param rows number of rows
+     * @param cols number of columns
+     * @return returns the number of islands
+     * Note: assumes non-empty tiles map
+     */
+    public static int countIslands(boolean[][] tilesMap, int rows, int cols){
+        boolean[][] visited = new boolean[rows][cols];
+        int numIslands = 0;
+        for(int i =0; i < rows; i++){
+            for (int j=0; j <rows; j++){
+                if(tilesMap[i][j] && !visited[i][j]){
+                    visited[i][j]= true;
+                    countIslandsHelper(tilesMap, visited, i, j);
+                    numIslands++;
+                }
+            }
+        }
+        return numIslands;
+    }
+
+    /**
+     * Recursive helper function for countIslands
+     * and iterate through the elements of the similar islands
+     * @param tiles 2d map of boolean tiles
+     * @param visited 2d map to see if a tile is already visited (boolean)
+     * @param i rows ith position in the tiles map to be checked
+     * @param j columns jth position in the tiles map to be checked
+     */
+    private static void countIslandsHelper(boolean[][] tiles, boolean [][]visited, int i, int j){
+        //check for its neighbors
+        for(Neighbour n: getNeighbours(i,j, tiles.length, tiles[0].length)){
+            if(tiles[n.row][n.col]  && !visited[n.row][n.col]){
+                visited[n.row][n.col]=true;
+                countIslandsHelper(tiles, visited,n.row, n.col);
+            }
+        }
+    }
+
+    /**
+     * Returns the horizontal and vertical neighbours of a selected tile
+     * @param row row index of the selected tile
+     * @param col column index of the selected tile
+     * @param rowUpperLimit number of rows in the tiles map
+     * @param colUpperLimit number of columns in the tiles map
+     */
+    private static HashSet<Neighbour> getNeighbours(int row, int col, int rowUpperLimit, int colUpperLimit){
```

Comment:
It would be cleaner in this particular case to actually just have four checks instead of a double loop with so many constraints